### PR TITLE
Bailout early from `isFunctionObjectType` for evolving arrays

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/44659

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -27044,6 +27044,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function isFunctionObjectType(type: ObjectType): boolean {
+        if (getObjectFlags(type) & ObjectFlags.EvolvingArray) {
```
Comment (Andarist) [inline src/compiler/checker.ts:27047]: `resolveStructuredTypeMembers` is still going to crash if an evolving array manages to reach it. I concluded that it's still better - at least in this case - to bail out early. IIUC, it avoid "fixing" this evolving type without a strong reason to do so - so the evolving array can continue evolving in certain situations.

**Comment by jakebailey** [comment]
@typescript-bot test it

File: src/compiler/checker.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/checker.ts]: Are there other paths whereby an evolving array could reach resolveStructuredTypeMembers? This seems like an important limitation that needs to be checked, either multiple places or maybe within resolveStructuredTypeMembers itself.

Comment (Andarist) [reply src/compiler/checker.ts]: > Are there other paths whereby an evolving array could reach resolveStructuredTypeMembers? 

None that we are aware of. Any call with an evolving array to that function would throw and I didn't see any other crashes reported there. It's of course not completely unlikely that there are other ways to end up with that crash.

> This seems like an important limitation that needs to be checked, either multiple places or maybe within resolveStructuredTypeMembers itself.

There is already an assert there. The only thing that potentially could be improved is to switch the assert to some other logic that would handle evolving arrays gracefully. I don't know what that logic would be and I think if this crash is a rare one then this fix is good enough.

**Review by andrewbranch — APPROVED** [review]

**Review by jakebailey — APPROVED** [review]

**Pull request merged**